File: 动态规划/maxSubarraySumCircular/maxSubarraySumCircular.py
#没有解出来
#环形数组最大和
#解法分为两种情况，一、最大和出现在数组中间，则采用普通的动态规划即可解答；二、最大和出现在首尾，则可以转化为求中间的最小和，可以看作双指针往中间推，直到出现负数为止
#特殊情况为最大和小于0，此时最大值为一个元素




class Solution:
    def maxSubarraySumCircular(self, nums: List[int]) -> int:
        maxVal,minVal = -99999,99999
        curMax,curMin = 0,0
        totalSum = 0
        for i in nums:
            totalSum += i
            curMax = max([curMax+i,i])
            curMin = min([curMin+i, i])
            maxVal = max([curMax, maxVal])
            minVal = min([curMin, minVal])
        return maxVal if maxVal < 0 else max([maxVal, totalSum-minVal])

File: 动态规划/maxSubarraySumCircular/maxSubarraySumCircular.c
int maxSubarraySumCircular(int* nums, int numsSize){
    int total=0,max=-99999,min=99999,curMax=0,curMin=0;
    for(int j=0;j<numsSize;j++){
        int i = nums[j];
        total+=i;
        curMax = curMax+i > i ? curMax+i : i;
        curMin = curMin+i < i ? curMin+i : i;
        max = curMax > max ? curMax : max;
        min = curMin < min ? curMin : min;
    }
    //printf("%d,%d", max, min);
    return max > 0 ? (total-min > max ? total-min : max) : max;
}